src/groth16/aggregate/verify.rs

```diff
@@ -0,0 +1,639 @@
+    // order to construct them in logn time.
+    challenges.reverse();
+    challenges_inv.reverse();
+    // Verify commitment keys wellformed
+    let fvkey = proof.tmipp.gipa.final_vkey;
+    let fwkey = proof.tmipp.gipa.final_wkey;
+    // KZG challenge point
+    let c = oracle!(
+        &challenges.first().unwrap(),
+        &fvkey.0,
+        &fvkey.1,
+        &fwkey.0,
+        &fwkey.1
+    );
+
+    // we take reference so they are able to be copied in the par! macro
+    let final_a = &proof.tmipp.gipa.final_a;
+    let final_b = &proof.tmipp.gipa.final_b;
+    let final_c = &proof.tmipp.gipa.final_c;
+    let final_r = &proof.tmipp.gipa.final_r;
+    let final_zab = &final_res.zab;
+    let final_tab = &final_res.tab;
+    let final_uab = &final_res.uab;
+    let final_tc = &final_res.tc;
+    let final_uc = &final_res.uc;
+
+    let now = Instant::now();
+    par! {
+        // check the opening proof for v
+        let vtuple = verify_kzg_opening_g2(
+            v_srs,
+            &fvkey,
+            &proof.tmipp.vkey_opening,
+            &challenges_inv,
+            &E::Fr::one(),
+            &c,
+        ),
+        // check the opening proof for w - note that w has been rescaled by $r^{-1}$
+        let wtuple = verify_kzg_opening_g1(
+            v_srs,
+            &fwkey,
+            &proof.tmipp.wkey_opening,
+            &challenges,
+            &r_shift.inverse().unwrap(),
+            &c,
+        ),
+        //
+        // We create a sequence of pairing tuple that we aggregate together at
+        // the end to perform only once the final exponentiation.
+        //
+        // TIPP
+        // z = e(A,B)
+        let check_z = PairingCheck::<E>::from_miller_inputs(&[(final_a, final_b)], final_zab),
+        //  final_aB.0 = T = e(A,v1)e(w1,B)
+        let check_ab0 = PairingCheck::<E>::from_miller_inputs(&[(final_a, &fvkey.0),(&fwkey.0, final_b)], final_tab),
+
+        //  final_aB.1 = U = e(A,v2)e(w2,B)
+        let check_ab1 = PairingCheck::<E>::from_miller_inputs(&[(final_a, &fvkey.1),(&fwkey.1, final_b)], final_uab),
+
+        // MIPP
+        // Verify base inner product commitment
+        // Z ==  c ^ r
+        let final_z =
+            inner_product::multiexponentiation::<E::G1Affine>(&[final_c.clone()],
+            &[final_r.clone()]),
+        // Check commiment correctness
+        // T = e(C,v1)
+        let check_t = PairingCheck::<E>::from_miller_inputs(&[(final_c,&fvkey.0)],final_tc),
+        // U = e(A,v2)
+        let check_u = PairingCheck::<E>::from_miller_inputs(&[(final_c,&fvkey.1)],final_uc)
+    };
+
+    debug!(
+        "TIPP verify: parallel checks before merge: {}ms",
+        now.elapsed().as_millis(),
+    );
+
+    let b = final_z == final_res.zc;
+    // only check that doesn't require pairing so we can give a tuple that will
+    // render the equation wrong in case it's false
+    if !b {
+        return PairingCheck::new_invalid();
+    }
```

Comment:
I do not think we should be hiding failure values inside of `PairingCheck`s like this. It's way too concrete and ends up entangling the verification logic, the algebra, and the program control flow in a way that is hard to reason about and will only get worse with time and refactoring. Although I don't think this currently gives incorrect results, I do think it makes analysis more difficult and certainly increases the likelihood of problems down the line. This code base and layered optimizations are complicated enough already — without conflating the correctness arguments of so many different logical layers into an aggregate depending on the coupled behavior of all its parts.

This is clever, but too clever.

If you want to keep it relatively simple, you could pass `Option<PairingCheck>` through the channels and map over the options when merging and verifying, etc. and treat `None` as failure. 



For one thing, this makes

Reply:
It's not a **failure** - it is an indication the prover have given invalid values, as much as the rest of the tuples returned. I'm simply passing through the PairingCheck to signal this, otherwise I'd have to do a special trait "Check" with two implementations of "MultiExpCheck" and "PairingCheck" - that doesn't sound very useful to me. But I re-iterate: it is **not** a failure of any kind.

Reply:
So `Option` is not what should be done here since the output of this function doesn't tell if verification was good or not, it only gives back a temporary result. In this case we just know the end result prematurely but that doesn't mean there has been a failure with the code (like impossible to get randomness or stg else).

Reply:
Just for clarity, I meant 'failure to verify the aggregated proof'. I thing we've moved past this. The reason I suggested `Option` rather than some kind of `Error` is exactly that it is an ordinary, valid value. But it also encodes information that isn't inherently a pairing check. I wouldn't want to specify this as adding an extra pairing check carrying this information then optimizing it along with the other pairing checks. And I wouldn't like to implement it in a way that isn't or can't be specified.